### ISSUE TITLE
An idea about skipping the ads UX

### ISSUE THREAD

**Issue by KOLANICH** [dest4/buffer-player#2]
>Skipping ads in a podcast is trivial in terms of user experience: it is like skipping a part of a song. Unfortunately, it is not as simple for radio. Broadcasts are linear and for live content, fast-forwarding is not an option!

In fact it is obvious.

1 detect boundaries of songs on other streams, specified by user as fallback ones.
2 record `n` songs from other streams. Recording may be optimized by keeping only songs of suitable length and using music fingerprinting to determine songs length ahead of time. To optimize UX you can try to encode song "mood" into a vector using rnn, and then average moods of all the songs of the current stream to get "mood" of the stream and then keep only the songs having "mood" close enough to stream mood.
3 when an ad is skipped, predict the length of sequence of ads for that stream based on previous history and start playing the song of the suitable length, starting recording the songs in the current stream
4 when playing recorded stream, increase playing speed (compensate frequency shift!) by tolerable amount in order to get rid of the difference between recorded songs and ads.
5 ads can be removed ahead of time if it s beneficial. Just predict appearance of ads in near future, start recording the current stream classified as non-ads and start playing a fallback song.

**Comment by dest4** [comment]
Your ideas are hard to implement. If easy for you, please do it and I will happily examine your pull request.

First question: how do you reliably detect boundaries of songs?

**Comment by KOLANICH** [comment]
>Your ideas are hard to implement. If easy for you, please do it and I will happily examine your pull request.

I have not claimed that it is easy to implement. You have asked some brainstorming, so I have posted the ideas immediately came into my mind, that's why I called them obvious, it doesn't men they are easy to implement. I'm not very familiar with the specifics (so feel free to correct me), so they may be junk, but a brainstorm is a brainstorm.

So ...

>First question: how do you reliably detect boundaries of songs?

Not sure if my experience is relevant here (I don't listen online radio, I listen the analog one), but:

1. Metadata may be available.
2. Starts and ends of songs usually have almost no speech, just music.
3. I guess different songs would have differrent fingerprints. So it may be possible to compute constantly fingerprints and track sufficiently rapid changes in them. So which time resolution (taking into account that if we cut song from the beginning and no speech is cut, it should not be annoying) do we need and can fingerprinting window be less than the resolution required?
4. Different songs often have different sets of singers and different singers sometimes can be distinguished by their voices. So, cut songs into intervals, separate speech from music, encode voice into a vector, look for rapid changes in voices' vectors.
5. Songs are often interleaved by speech of a DJ.
